File: questions/54.SpiralMatrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
const spiralOrder = function (matrix) {
  const result = []
  let left = 0
  let right = matrix[0].length
  let top = 0
  let bottom = matrix.length

  while (left < right && top < bottom) {
    // top row
    for (let i = left; i < right; i++) {
      result.push(matrix[top][i])
    }
    top++

    // right column
    for (let i = top; i < bottom; i++) {
      result.push(matrix[i][right - 1])
    }
    right--

    if (!(left < right && top < bottom)) break

    // right to left in bottom row
    for (let i = right - 1; i >= left; i--) {
      result.push(matrix[bottom - 1][i])
    }
    bottom--

    // bottom to top in left
    for (let i = bottom - 1; i >= top; i--) {
      result.push(matrix[i][left])
    }
    left++
  }

  return result
}
console.log(
  spiralOrder([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
  ]),
)

File: leetcode/54.SpiralMatrix.go
package leetcode

func spiralOrder(matrix [][]int) []int {
	var result []int
	left := 0
	right := len(matrix[0])
	top := 0
	bottom := len(matrix)

	for left < right && top < bottom {
		// right to left in top row
		for i := left; i < right; i++ {
			result = append(result, matrix[top][i])
		}
		top++

		// top to bottom in right column
		for i := top; i < bottom; i++ {
			result = append(result, matrix[i][right-1])
		}
		right--

		// break if it reaches to the end
		if !(left < right && top < bottom) {
			break
		}

		// right to left in bottom row
		for i := right - 1; i >= left; i-- {
			result = append(result, matrix[bottom-1][i])
		}
		bottom--

		// bottom to top in left column
		for i := bottom - 1; i >= top; i-- {
			result = append(result, matrix[i][left])
		}
		left++
	}

	return result
}
